File: examples/ch04/SplitLines.hs
{-- snippet splitLines --}
splitLines :: String -> [String]

splitLines [] = []
splitLines cs = case break isLineSeparator cs of 
                  (pre, '\r':'\n':suf) -> pre : splitLines suf
                  (pre, '\r':suf) -> pre : splitLines suf
                  (pre, '\n':suf) -> pre : splitLines suf
                  (pre, "") -> [pre]

isLineSeparator :: Char -> Bool
isLineSeparator c = c `elem` "\r\n"
{-- /snippet splitLines --}

File: examples/ch04/splitlines.py
## snippet splitlines
def splitlines(s):
    ret = []
    while True:
        head, sep, tail = s.partition('\r\n')
        if not (sep or tail):
            head, sep, tail = s.partition('\r')
        if not (sep or tail):
            head, sep, tail = s.partition('\n')
        if not (sep or tail):
            break
        ret.append(head)
        s = tail
    if s:
        ret.append(s)
    return ret
## /snippet splitlines
